### PR TITLE
Allow install

### PR DESCRIPTION
**Pull request by SBalslev**
Always allow installation of app source apps for delegated admins.

Fixes [AB#551407](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/551407)

### CODE REVIEW

**Review by StefanSosic — APPROVED** [review]

**Review by JesperSchulz — DISMISSED** [review]

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 1]
```diff
@@ -96,12 +96,16 @@ codeunit 2515 "AppSource Product Manager"
...
     var
+        AzureADUserManagement: Codeunit "Azure AD User Management";
         PlanToken: JsonToken;
         PlanObject: JsonObject;
         PricingTypesToken: JsonToken;
         PricingTypes: JsonArray;
         PricingType: JsonToken;
     begin
+        if (AzureADUserManagement.IsUserDelegated(Database.UserSecurityId())) then
+            exit(true); // Delgated admins are always allowed to install.
```
Comment (zeande) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: Super minor: Spelling
```suggestion
            exit(true); // Delegated admins are always allowed to install.
```

**Review by zeande — APPROVED** [review]

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 2]
```diff
@@ -96,12 +96,16 @@ codeunit 2515 "AppSource Product Manager"
...
     internal procedure CanInstallProductWithPlans(Plans: JsonArray): Boolean
     var
+        AzureADUserManagement: Codeunit "Azure AD User Management";
         PlanToken: JsonToken;
         PlanObject: JsonObject;
         PricingTypesToken: JsonToken;
         PricingTypes: JsonArray;
         PricingType: JsonToken;
     begin
+        if (AzureADUserManagement.IsUserDelegated(Database.UserSecurityId())) then
```
Comment (mortenseifert) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: Not really a bug - but in AL we would not encapsulate a single condition in appetencies.  

```suggestion
        if AzureADUserManagement.IsUserDelegated(Database.UserSecurityId()) then
```

**Review by mortenseifert — CHANGES_REQUESTED** [review]

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al (hunk 2 above)
Comment (JesperSchulz) [reply src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: I wonder why this isn't a CodeCop rule - I actually thought it was.

File: src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al [hunk 3]
```diff
@@ -178,11 +178,18 @@ page 2516 "AppSource Product Details"
 
                 trigger OnAction()
                 var
+                    AzureADUserManagement: Codeunit "Azure AD User Management";
```
Comment (StefanSosic) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al:182]: Now failing due missing usings

**Review by StefanSosic — CHANGES_REQUESTED** [review]

File: src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al [hunk 4]
```diff
@@ -231,6 +238,7 @@ page 2516 "AppSource Product Details"
         PlansOverview: Text;
         PlansAreVisible: Boolean;
         PurchaseLicensesElsewhereLbl: Label 'Installing this app might lead to undesired behavior if licenses are not purchased before use. You must purchase licenses through Microsoft AppSource.\Do you want to continue with the installation?';
+        InstallNewAppWarningLbl: Label 'You are installing a new AppSource app.\If this is a Contact Me app, please ensure that you have contacted the publisher before proceeding to ensure you have the proper licensing, guidance and app dependencies in place. If this is a transactable plan, you will need to have purchased appropriate licenses on the online AppSource Marketplace to be able to use the app.\Do you want to continue with the installation?';
```
Comment (pborring) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al]:
```suggestion
        InstallNewAppWarningLbl: Label 'You are installing a new AppSource app.\If this is a Contact Me app, please ensure that you have contacted the publisher before proceeding to ensure you have the proper licensing, guidance and app dependencies in place. If this is a transactable plan, the customer must purchase appropriate licenses on the online AppSource Marketplace to be able to use the app.\Do you want to continue with the installation?';
```

File: src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al [hunk 5]
```diff
@@ -231,6 +239,7 @@ page 2516 "AppSource Product Details"
         PlansOverview: Text;
         PlansAreVisible: Boolean;
         PurchaseLicensesElsewhereLbl: Label 'Installing this app might lead to undesired behavior if licenses are not purchased before use. You must purchase licenses through Microsoft AppSource.\Do you want to continue with the installation?';
+        InstallNewAppWarningLbl: Label 'You''re installing a new AppSource app.\If this is a Contact Me app, make sure that you have contacted the publisher before proceeding to ensure that you have the proper licensing, guidance, and app dependencies in place. If this is a transactable plan, you will need to have purchased appropriate licenses on the online AppSource Marketplace to be able to use the app.\Do you want to continue with the installation?';
```
Comment (pborring) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al]:
```suggestion
        InstallNewAppWarningLbl: Label 'You''re installing a new AppSource app.\If this is a Contact Me app, make sure that you have contacted the publisher before proceeding to ensure that you have the proper licensing, guidance, and app dependencies in place. If this is a transactable plan, the customer will need to purchase appropriate licenses on Microsoft AppSource to be able to use the app.\Do you want to continue with the installation?';
```

File: src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al [hunk 6]
```diff
@@ -178,11 +179,18 @@ page 2516 "AppSource Product Details"
...
                     ExtensionManagement: Codeunit "Extension Management";
                 begin
-                    if (PlansAreVisible) then
-                        if not Confirm(PurchaseLicensesElsewhereLbl) then
+                    // Delegated admins are always allowed to install so they get a different warning
+                    if AzureADUserManagement.IsUserDelegated(Database.UserSecurityId()) then begin
+                        if not Confirm(InstallNewAppWarningLbl) then
                             exit;
+                    end else
+                        if (PlansAreVisible) then
```
Comment (JesperSchulz) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al]: You usually don't do this in AL... The if (stmt). Drop the parenthesis.

**Review by JesperSchulz — DISMISSED** [review]

**Review by JesperSchulz — APPROVED** [review]

**Review by StefanSosic — APPROVED** [review]

**Review by blrobl — APPROVED** [review]

**Review by mortenseifert — APPROVED** [review]

**Pull request merged**